src/tigris.ts

```diff
@@ -16,10 +16,59 @@ import {
 	ServerMetadata
 } from "./types";
 
+import {GetAccessTokenRequest as ProtoGetAccessTokenRequest} from "./proto/server/v1/auth_pb";
+
 import {DB} from "./db";
+import {AuthClient} from "./proto/server/v1/auth_grpc_pb";
+import {Utility} from "./utility";
+
+const AuthorizationHeaderName = "authorization";
+const AuthorizationBearer = "Bearer ";
 
 export interface TigrisClientConfig {
 	serverUrl: string;
+	insecureChannel?: boolean
+	refreshToken?: string
+}
+
+class TokenSupplier {
+	private refreshToken: string;
+	private accessToken: string;
+	private nextRefreshTime: number;
+	private authClient: AuthClient;
+
+	constructor(config: TigrisClientConfig) {
+		this.authClient = new AuthClient(config.serverUrl, grpc.credentials.createSsl());
+	}
+
+	getAccessToken(): string {
+		if (this.shouldRefresh()) {
+			// refresh
+			this.authClient.getAccessToken(
+				new ProtoGetAccessTokenRequest().setRefreshToken(this.refreshToken),
+				(error, response) => {
+					if (error) {
+						// log failure
+					} else {
+						this.accessToken = response.getAccessToken();
+						this.refreshToken = response.getAccessToken();
+						// retrieve exp
+						const parts: string[] = this.accessToken.split("\\.");
+						const exp = Number(Utility.jsonStringToObj(Utility._base64Decode(parts[1]))["exp"]);
+						// 5 min before expiry (note: exp is in seconds)
+						this.nextRefreshTime = (exp * 1000) - 300_000;
```

Comment:
We should probably jitter this

Reply:
expiration will depend on first manual login action. So not all refresh will be concentrated at the same time. Is there any specific reason we should add jitter here?

Reply:
We don't have explicit control over these login actions (uncontrolled tigris restart, client namespace generating 100s of connections for all to recycle at the same time). Unless there is tight SLA of 300 seconds we can window this to 200-300 or 250-350 etc.

Not needed now but could easily slip under the nose.

Reply:
good point for restarts, added 1-5min of jitter here.